File: goblinframework-api/src/main/java/org/goblinframework/api/annotation/ServiceVersion.java
package org.goblinframework.api.annotation;

import java.lang.annotation.*;

@Documented
@Target({ElementType.TYPE, ElementType.METHOD, ElementType.FIELD})
@Retention(RetentionPolicy.RUNTIME)
public @interface ServiceVersion {

  String version() default "1.0";

  boolean enable() default true;

}

File: goblin-rpc/src/main/java/org/goblinframework/rpc/service/_ServiceVersion.kt
package org.goblinframework.rpc.service

import org.goblinframework.api.rpc.ImportService
import org.goblinframework.api.rpc.ServiceVersion

fun calculateServerVersion(annotation: ImportService?): String? {
  return annotation?.run {
    if (this.enable) {
      calculateServerVersion(this.version)
    } else {
      null
    }
  }
}

fun calculateServerVersion(annotation: ServiceVersion?): String? {
  return annotation?.run {
    if (this.enable) {
      val s: String = this.version
      if (s.isBlank()) null else s.trim()
    } else {
      null
    }
  }
}